src/rpm/builder.rs

```diff
@@ -832,21 +829,26 @@ impl PackageBuilder {
             base_names.push(entry.base_name.to_owned());
             // @todo: is there a use case for not performing all verifications? and are we performing those verifications currently anyway?
             file_verify_flags.push(entry.verify_flags.bits());
-            let content = entry.content.to_owned();
             if !uses_large_files {
-                let mut writer = payload::Builder::new(cpio_path)
+                let writer = payload::Builder::new(cpio_path)
                     .mode(entry.mode.into())
                     .ino(ino_index)
                     .uid(self.uid.unwrap_or(0))
                     .gid(self.gid.unwrap_or(0))
-                    .write_cpio(&mut archive, content.len() as u32);
-                writer.write_all(&content)?;
-                writer.finish()?;
+                    .write_cpio(&mut archive, entry.source.size()? as u32);
+                let mut hash_writer = HashAndWrite::new(Sha256::new(), writer);
+                io::copy(&mut entry.source.buf_read()?, &mut hash_writer)?;
+                let (hash_output, cpio_writer) = hash_writer.finalize();
+                file_hashes.push(hex::encode(hash_output));
+                cpio_writer.finish()?;
             } else {
                 // @todo: can we just use ino_index instead of file_index?
                 let header = payload::stripped_cpio_header(file_index as u32);
                 archive.write_all(&header)?;
-                archive.write_all(&content)?;
+                let mut hash_writer = HashAndWrite::new(Sha256::new(), &mut archive);
+                io::copy(&mut entry.source.buf_read()?, &mut hash_writer)?;
+                let (hash_output, _) = hash_writer.finalize();
+                file_hashes.push(hex::encode(hash_output));
                 archive.flush()?;
```

Comment:
How about returning `hash_output` from this block so that the duplication is eliminated?